### PR TITLE
Generate lang.json and OCaml code from Python/mypy source

### PR DESCRIPTION
**Pull request by mjambon**
The old setup consisted of a handwritten `lang.json` file and a readme.

The new setup uses a typed Python program that guarantees the well-formedness of the language data and allows for comments where needed. It generates two kinds of files:
* `lang.json` which remains compatible with the previous version of this file.
* `Language.mli` and `Language.ml` which can be used directly in the OCaml project (semgrep-core/osemgrep) without introducing an unusual build step to the project. Removing this step was the primary motivation for this rewrite.

Output for other languages could also be added if people find it more convenient than using the JSON file.

Instructions for maintaining this are in the readme and in the `generate` file.

### CODE REVIEW

**Comment by aryx** [comment]
Why not use ATD to specify the schema instead of those python classes, and then use atdpy to generate the conversion boilerplate and then use Python to define the 'list' using the class generated by atdpy and then generate the .json
and Language.ml from it? So leverage ATD as much as possible, and use python for 'list' which you can't really define in ATD.

Also maybe it would be better to move this inside semgrep-interfaces/. I think semgrep-langs was created when there was no semgrep-interfaces repo yet, but now that we have one, better to centralize everything in one place.

**Comment by mjambon** [comment]
Initially, I wanted to use ATD because it defines a type (better than the readme). I tried that 6 months ago or so but didn't continue the effort because of the complexity.

We need the following:

1. Generate type definitions for OCaml, Python, etc. that correspond to each language.
2. We need to take the list of languages (data) and turn it into code that OCaml, Python, TypeScript, etc. can use.

Step 1 can be done with ATD well as we've done before in other contexts. Step 2 is done in Python (pysemgrep) by loading a JSON file at runtime. For OCaml (semgrep-core), it was done by an extra build step that takes the JSON data and turns it into OCaml (using jinja). I wanted to eliminate the jinja build step which is weird to have in an OCaml project (extra setup, needs extra expertise). Dune also buries the generated OCaml files (same problem with atdgen btw), so we end up without easy access to the contents of Lang.mli.

To make the list of languages more accessible, the only approach I could see is to generate the OCaml files ahead of time, in a step outside of the dune build. This could be done either in the semgrep-langs repo or in the semgrep repo. Either would be fine as far as OCaml is concerned.

However, we also need to produce a `lang.json` or equivalent Python and TypeScript code that contains the list of languages. Doing this with ATD involves (a) generating the target code in OCaml, Python, and TypeScript, and (b) running a program that uses that code and defines the list of languages. That program could be written in OCaml but it requires an opam setup. It could be written in Python, which is more likely to be installed on the user's machine. That's the solution I went for, thinking nobody would complain that we depend on a bunch of Python packages (in fact, the dependencies are rather lightweight and don't involve 3rd-party packages like atdgen-runtime for OCaml).

Now, given the solution of writing a Python program to generate `lang.json`, we might as well define the type definitions (which are about 15 lines of code) in Python/mypy directly rather than ATD and duplicate it manually for OCaml. This is what's being done right now. It avoids the atdpy step and introduces a tiny bit of duplication of type definitions (somewhere, we `print()` an OCaml type definition similar to the mypy type definition of a language). Writing `lang.json` is the other part of the code where atdpy-generated Python code would have saved a bit of boilerplate. atdgen doesn't provide a way to write data in OCaml format like we wanted without going through JSON, so that's not an effort we could avoid.

In conclusion, having a single `generate` script that only depends on a vanilla installation of Python 3 is simpler than ATD to set up for everyone who will have to touch this. ATD wouldn't make it easier to generate the list of languages in OCaml syntax, and Python is fine for this task. If the atd suite provided a direct way to generate data for all the target languages instead of going through JSON, we would certainly use it.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Generate code representing the list of target languages supported
+# by Semgrep.
+#
+
+.PHONY: build
+build:
+	mypy generate
+	./generate
```
Comment (tpetr) [inline Makefile:8]: uber-nit: can we name this file `generate.py` so that IDEs will do the right thing?

File: generate [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+
+    # Extension used when generating temporary or downloable files from
+    # code snippets, hinting at what the file contains.
+    example_ext: Optional[str] = None
+
+    ##################################################################
+    # Properties used to guess the language of a file.
+    ##################################################################
+    # This is useful for guessing purposes.
+    # Hoewver, these properties are too simplistic to fully determine
```
Comment (tpetr) [inline generate:61]:
```suggestion
    # However, these properties are too simplistic to fully determine
```

File: generate [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+    # target filtering as done by 'semgrep scan'.
+
+    # File extensions that are commonly used for the language, including
+    # the period.
+    # Several languages may share an extension.
+    # TODO: document what qualifies as an extension since don't seem to
+    # they have to start with a ".".
+    # TODO: document if they're case-sensitive.
+    exts: List[str] = field(default_factory=lambda: [""])
```
Comment (tpetr) [inline generate:70]: the zero state of `exts` surprised me at first, what do you think about adding a clarifying comment here?

also, unrelated, what happens if a language has an empty list for `exts`?

**Review by tpetr — APPROVED** [review]
Changes look good to me, just some nits.

+1 to @aryx's question about consolidating things in semgrep-interfaces

Also, we may want to add a pre-commit config to ensure that the ocaml and JSON files are consistent with the definitions in `generate`

**Comment by aryx** [comment]
Ok, I like your explanation martin. Would be nice to add it somewhere in the python code, like in a comment
in the generate script: "why those python defs below instead of using ATD:" and then your explanations.

I like to remember design decisions in comments in the relevant code so that in 6 months I can remember
without asking why we didn't use ATD here.

File: Makefile (hunk 1 above)
Comment (mjambon) [reply Makefile:8]: I despise this practice. It's a command, it should not have an extension. How about a symlink? I'll make a symlink, it will make me happy.

File: generate (hunk 3 above)
Comment (mjambon) [reply generate]: I don't know why `""` is accepted as an extension. I also don't know why `[""]` is required when a language doesn't have specific extensions. I could change it to the empty list `[]` and fix the code in semgrep but I don't know what this would break elsewhere.

Comment (mjambon) [reply generate]: It's going to be just a comment for now. I might fix this later.

File: Makefile (hunk 1 above)
Comment (mjambon) [reply Makefile:8]: added a symlink
